azurerm/internal/services/compute/linux_virtual_machine_scale_set_resource.go

```diff
@@ -702,13 +702,23 @@ func resourceLinuxVirtualMachineScaleSetUpdate(d *schema.ResourceData, meta inte
 		}
 
 		if d.HasChange("source_image_id") || d.HasChange("source_image_reference") {
+			// Must include all storage profile properties when updating disk image.  See: https://github.com/terraform-providers/terraform-provider-azurerm/issues/8273
+			ultraSSDEnabled := d.Get("additional_capabilities.0.ultra_ssd_enabled").(bool)
+			dataDisks, _ := ExpandVirtualMachineScaleSetDataDisk(d.Get("data_disk").([]interface{}), ultraSSDEnabled)
+			if err != nil {
+				return fmt.Errorf("expanding `data_disk`: %+v", err)
+			}
+
+			osDiskRaw := d.Get("os_disk").([]interface{})
+
 			sourceImageReferenceRaw := d.Get("source_image_reference").([]interface{})
 			sourceImageId := d.Get("source_image_id").(string)
 			sourceImageReference, err := expandSourceImageReference(sourceImageReferenceRaw, sourceImageId)
 			if err != nil {
 				return err
 			}
-
+			updateProps.VirtualMachineProfile.StorageProfile.DataDisks = dataDisks
+			updateProps.VirtualMachineProfile.StorageProfile.OsDisk = ExpandVirtualMachineScaleSetOSDiskUpdate(osDiskRaw)
 			updateProps.VirtualMachineProfile.StorageProfile.ImageReference = sourceImageReference
```

Comment:
rather than pulling this from the config, since it's possible `ignore_changes` could be being used here, we should likely use the value from the API here instead, which is available via:

> existing.VirtualMachineScaleSetProperties.VirtualMachineProfile.StorageProfile.OSDisk
> existing.VirtualMachineScaleSetProperties.VirtualMachineProfile.StorageProfile.DataDisks

Reply:
Thanks @tombuildsstuff -- I've incorporated the changes you requested.  Let me know if there's anything else I can add here. 👍